File: rust/examples/interop_echo_client.rs
use std::env;

use forpc::RpcPeer;
use forpc::pb::test::{EchoRequest, EchoResponse};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let mut args = env::args().skip(1);
    let url = args
        .next()
        .unwrap_or_else(|| "tcp://127.0.0.1:24000".to_string());
    let msg = args.next().unwrap_or_else(|| "Hello".to_string());

    let peer = RpcPeer::connect_with_retry(&url, 50).await?;

    let peer_clone = peer.clone();
    tokio::spawn(async move {
        let _ = peer_clone.serve().await;
    });

    let resp: EchoResponse = peer
        .call(
            "Test/Echo",
            EchoRequest { data: msg },
        )
        .await?;

    println!("reply: {}", resp.result);
    Ok(())
}

File: node/scripts/interop_echo_client.js
const { Peer } = require('..')
const protobuf = require('protobufjs')
const path = require('path')

const url = process.argv[2] || 'tcp://127.0.0.1:24000'
const msg = process.argv[3] || 'Hello'

;(async () => {
  const root = await protobuf.load(path.join(__dirname, '../../proto/forpc_test.proto'))
  const EchoRequest = root.lookupType('forpc.test.EchoRequest')
  const EchoResponse = root.lookupType('forpc.test.EchoResponse')

  const p = await Peer.connect(url)
  const reqBuf = Buffer.from(EchoRequest.encode(EchoRequest.create({ data: msg })).finish())
  const respBuf = await p.callRaw('Test/Echo', reqBuf)
  const resp = EchoResponse.decode(respBuf)
  process.stdout.write(`reply: ${resp.result}\n`)
})().catch((e) => {
  console.error(e)
  process.exit(1)
})
